crates/environ/src/compile/module_environ.rs

```diff
@@ -949,30 +936,54 @@ impl ModuleTranslation<'_> {
                 segments: Vec::new(),
             });
         }
-        let mut idx = 0;
-        let ok = self.module.memory_initialization.init_memory(
-            &mut (),
-            InitMemory::CompileTime(&self.module),
-            |(), memory, init| {
+
+        struct InitMemoryCompile<'a> {
+            module: &'a Module,
+            info: &'a mut PrimaryMap<MemoryIndex, Memory>,
+            idx: usize,
+        }
+        impl InitMemory for InitMemoryCompile<'_> {
+            fn memory_size_in_pages(&mut self, memory_index: MemoryIndex) -> u64 {
+                self.module.memory_plans[memory_index].memory.minimum
+            }
+
+            fn eval_offset(&mut self, memory_index: MemoryIndex, expr: &ConstExpr) -> Option<u64> {
+                let mem64 = self.module.memory_plans[memory_index].memory.memory64;
+                match expr.ops() {
+                    &[ConstOp::I32Const(offset)] if !mem64 => Some(offset.unsigned().into()),
+                    &[ConstOp::I64Const(offset)] if mem64 => Some(offset.unsigned()),
+                    _ => None,
+                }
+            }
```

Comment:
It might be better to turn `ConstEvalContext` into a trait -- I initially did that but then every use site was the same so I made it concrete. But if it were a trait, then we could use a `Result` for getting globals and funcrefs and return an `Err` when they aren't available at compile time. This would allow both reusing `eval` here so you don't have to do this matching (nor in the other PR, iirc) and also support const expression offsets like `(i32.add (i32.const 3) (i32.const 5))` for compile-time heap image computation.

Reply:
It turns out it's easy to turn `ConstEvalContext` into a trait, but it's not useful without bigger changes that I haven't been able to figure out yet.

`ConstExprEvaluator` can only be used within the wasmtime crate, because it relies on `ValRaw`. Everything that does const evaluation in that crate already has access to an `Instance`. All the places that need const evaluation before instantiation are in wasmtime-environ or wasmtime-types.

I tried factoring out a type-specific, integer-only const-evaluator that can live in wasmtime-types alongside `ConstExpr` and `ConstOp`. But the general-purpose const-evaluator is still necessary for initializing globals and I'm not real happy with duplicating this functionality.

Maybe we need a trait in wasmtime-types that `Val` or `ValRaw` can implement?

Reply:
> Maybe we need a trait in wasmtime-types that `Val` or `ValRaw` can implement?

Yeah I think we'll have to do something roughly like this.